core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1105,7 +1105,16 @@ private[spark] class DAGScheduler(
   private def submitMissingTasks(stage: Stage, jobId: Int) {
     logDebug("submitMissingTasks(" + stage + ")")
 
-    // First figure out the indexes of partition ids to compute.
+    // Before find missing partition, do the intermediate state clean work first.
+    // The operation here can make sure for the intermediate stage, `findMissingPartitions()`
+    // returns all partitions every time.
+    stage match {
```

Comment:
That's for the scenario of ExecutorLost. While executor lost happened, there's possible for the indeterminate stage rerun triggered by `submitParentStage`.
So if we only unregister during failure handling, only fetch failed stage and its parent stage do unregister, that logic would not cover the scenario of its parent's parent stage is indeterminate and have missing tasks.

Reply:
It seems we can still put the unregister logic into the block:
https://github.com/apache/spark/pull/25620/files#diff-6a9ff7fb74fd490a50462d45db2d5e11R1626 ?

Reply:
Actually the place points out by Xingbo is just my first-time attempt :), I found that's not enough to fix this problem during doing the integrate test.
This is because the logic we calculate `stagesToRollback` in `collectStagesToRollback`, only care about the downstream stages of the current fetch failed stage. For upstream indeterminate stages, put the unregister logic in failure handling didn't cover. So the correctness bug will still happen.

Reply:
I'm a little concerned about putting this here, as you'll see lower down in this method there is some handling for the case that `submitMissingTasks` is called but there are actually no tasks to run.  I'm not seeing how that happens now, but your change would make those cases always re-evaluate all partitions of the stage.

I think @jiangxb1987 suggestion makes sense, couldn't you do it the unregistering there?  I agree the logic is currently insufficient as its not building up the full set of stages that need to be recomputed, but maybe we need to combine both.

or maybe we understand the old cases of submitting a stage with no missing partitions and my concern is not relevant?

Reply:
Thanks for the detailed comment.
I tried to combine both at first but seems kind of duplicated code.
For the concerned about no missing partitions stage submission, I checked`shuffleMapStage.isAvailable` for the unregister in aa3a409, what we need here is making sure the partially completed indeterminate stage will be whole stage rerun.